expression/new_constant_propagation.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"bytes"
+
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	log "github.com/sirupsen/logrus"
+)
+
+// exprSet is a Set container for expressions, each expression in it is unique.
+// `tombstone` is deleted mark, if tombstone[i] is true, data[i] is invalid.
+// `index` use expr.HashCode() as key, to implement the unique property.
+type exprSet struct {
+	data       []Expression
+	tombstone  []bool
+	index      map[string]struct{}
```

Comment:
why not using `StringSet`?

Reply:
Because the data it stored is `[]Expression` @zz-jason

Reply:
No, the data it stores is the hash code, which is a string.

Reply:
`index` is the index of the `data`, and the data is `[]Expression`

Reply:
I'm a bit confused, do you mean change `exprSet` to `StringSet`  or change the name of `index` ?

Reply:
I mean, change `index      map[string]struct{}` to `exists set.StringSet`